csharp/src/Ice/ConnectionI.cs

```diff
@@ -1447,24 +1450,57 @@ internal void sendHeartbeat()
         {
             if (isActiveOrHolding())
             {
-                OutputStream os = new OutputStream(_instance, Util.currentProtocolEncoding);
-                os.writeBlob(Protocol.magic);
-                Util.currentProtocol.ice_writeMembers(os);
-                Util.currentProtocolEncoding.ice_writeMembers(os);
-                os.writeByte(Protocol.validateConnectionMsg);
-                os.writeByte(0);
-                os.writeInt(Protocol.headerSize); // Message size.
-                try
-                {
-                    sendMessage(new OutgoingMessage(os, false, false));
+                // We check if the connection has become inactive.
+                if (
+                    _inactivityTimer is null &&           // timer not already scheduled
+                    _inactivityTimeout > TimeSpan.Zero && // inactivity timeout is enabled
+                    _state == StateActive &&              // only schedule the timer if the connection is active
+                    _dispatchCount == 0 &&                // no pending dispatch
+                    _asyncRequests.Count == 0 &&          // no pending invocation
+                    _readHeader &&                        // we're not waiting for the remainder of an incoming message
+                    _sendStreams.Count <= 1)              // there is at most one pending outgoing message
+                {
+                    // We may become inactive while the peer is back-pressuring us. In this case, we only schedule the
+                    // inactivity timer if there is no pending outgoing message or the pending outgoing message is a
+                    // heartbeat.
+
+                    // The stream of _sendStreams.First is in _writeStream.
+                    if (_sendStreams.Count == 0 || isHeartbeat(_writeStream))
```

Comment:
When does ` isHeartbeat(_writeStream)` return `True` and `_inactivityTimer` is `null`?

Reply:
The two conditions are independent:
 - _inactivityTimer is null means the connection is active at the RPC level - for example there is an outstanding request, or an outstanding dispatch
 - isHeartbeat(_writeStream) == true means we're currently writing a heartbeat. This is an unusual situation since heartbeats are usually sent synchronously.  The peer is most likely back-pressuring us when this happens, or there is a transport issue.

Reply:
Say we are sending a heartbeat and `_sendStreams.Count == 0`, we schedule the inactivity timer and queue the heartbeat for sending.

If the peer is back pressuring us the heartbeat remains in `_writeStream`. 

When we are about to send the next heartbeat, the previous heartbeat remains in `_writeStream`, and we reschedule the inactivity timer.

I don't understand why the `isHeartbeat(_writeStream)` check is important here.